File: harry/stage6/Deal.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

public class Deal {

    Queue<Command> commandQueue = new LinkedList<>();
    Queue<Command> rollbackQueue = new LinkedList<>();

    public Deal() {
    }

    public void setCommand(Command command) {
        commandQueue.add(command);
    }

    public void play() {
        rollbackQueue.add(commandQueue.peek());
        // 로직 실행
        try {
            while (!commandQueue.isEmpty()) {
                commandQueue.poll().execute();
            }
        } catch (Exception e) {
            // 에러 발생한 경우 보상로직 실행~
            e.printStackTrace();
            while (!rollbackQueue.isEmpty()) {
                rollback();
            }
        }

    }

    public void rollback() {
        try {
            rollbackQueue.poll().undo();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: harry/stage6/ktVer/Deal.kt
package ktVer

import java.util.*

class Deal {
    val commandQueue: Queue<Command> = LinkedList<Command>()
    val rollbackQueue: Queue<Command> = LinkedList<Command>()

    fun setCommand(command: Command) = commandQueue.add(command)

    fun play(): Unit {
        try {
            while (!commandQueue.isEmpty()) {
                val rollbackCommandTemp = commandQueue.peek()
                commandQueue.poll().execute()
                rollbackQueue.add(rollbackCommandTemp)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            while (!rollbackQueue.isEmpty()) {
                rollback()
            }
        }
    }

    private fun rollback(): Unit {
        try {
            rollbackQueue.poll().undo()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

}